functions/index.js

```diff
@@ -95,85 +95,134 @@ exports.TrackUsers = functions.https.onRequest(async (req, res) => {
+  // and for course objects that are not in fetchedCoursesSoFar
+  const filteredCourses = filteredAllCourses[roster].filter(
+    rosterCourse =>
+      crseIds.includes(rosterCourse.crseId) &&
+      !fetchedCoursesSoFar.some(
+        course =>
+          course.subject === rosterCourse.subject && course.catalogNbr === rosterCourse.catalogNbr
+      )
+  );
+
+  // Update fetchedCourses with filteredCourses
+  filteredCourses.forEach(filteredCourse => {
+    filteredCourse.roster = roster; // Manually add roster field
+
+    let filteredCrseId = filteredCourse.crseId;
+
+    // Remove already fetched course ids
+    crseIds = crseIds.filter(crseId => crseId != filteredCrseId);
+
+    fetchedCourses.push(filteredCourse);
+  });
+
+  // Log courses that could not be fetched
+  crseIds.map(a => logUnfetchedCourse(a, roster));
+
+  return fetchedCourses;
 }
 
-/** FetchCourses fetches the most recent course objects for the list of
- * courseCodes in its input data object.
+/** FetchCourses fetches the course objects for the
+ * crseInfo in its input data object.
  *
- * In order to be a valid request, there must be a courseCodes property that is
- * a list of course code strings (e.g. 'CS 1110').
+ * An example of a data object:
+ * {"data": {"crseInfo": [{"roster": "SP20", "crseCodes": ["CS 1110"],
+ * "crseIds": [358578]}], "allowSameCourseForDifferentRosters": true}}
+ *
+ * In this case we want to get the course objects that correspond to the "CS 1110"
+ * crseCode for the "SP20" roster and all the course objects that correspond to the
+ * crseId 358578 for the "SP20" roster. Since allowSameCourseForDifferentRosters
+ * is set to true, we are allowing for the same courses in different rosters to be fetched.
+ *
+ * In order to be a valid request, there must be a crseInfo property that is a
+ * list of objects with the roster prop. crseCodes and crseIds are optional.
+ * Also, there must be an allowSameCourseForDifferentRosters property that is
+ * a boolean.
  */
 exports.FetchCourses = functions.https.onCall(data => {
-  let courseCodes = data.courseCodes.map(a => a.toUpperCase());
+  // Total accumulator of fetched courses
   let fetchedCourses = [];
 
-  // Sort rosters from most recent to least recent
-  let rosters = sortByMostRecentRosters(Object.keys(filteredAllCourses));
-  for (let i = 0; i < rosters.length && courseCodes.length > 0; i += 1) {
-    const roster = rosters[i];
-    let allRosterCourses = filteredAllCourses[roster];
-    // Filter for course objects whose code is in courseCodes
-    let filteredCourses = allRosterCourses.filter(
-      rosterCourse =>
-        courseCodes.indexOf(rosterCourse.subject.concat(' ', rosterCourse.catalogNbr)) != -1
+  // Iterate over each dataObject in data.crseInfo
+  // {"roster": "SP20", "crseCodes": ["CS 1110"], "crseIds": [358578]} is an example
+  // of a dataObject
+  data.crseInfo.forEach(dataObject => {
+    let roster = dataObject.roster;
```

Comment:
Now that the roster is not generated by this function itself but passed in as input, do we want to error out if the roster is not valid?

Reply:
That's definitely a good point. Should we console log an invalid roster? In the case where we have:
```
data.crseInfo === [{roster: 'sklfjaklf', crseCodes: ['CS 1110', 'INFO 2300']}, {roster: 'FA20', crseCodes:['CS 3410']}]
```
 erroring out on the first invalid roster would prevent the other valid course object of roster 'FA20' from being fetched.